### PR TITLE
refactor(dyno): convert all asserts to custom implementation

### PR DESCRIPTION
**Pull request by arezaii**
This PR implements dynamic evaluation of assertions in the dyno
compiler library code. Previously, we relied on standard `assert`s,
but desired a mechanism to dynamically control when they fired, and
one to control if failed asserts were fatal or if compilation could
attempt to continue.

With these changes, the compiler can be built with `CHPL_DEVELOPER=0`
(or unset), using `make ASSERTS=0` and all dyno asserts will be folded
out. If building with `CHPL_DEVELOPER=1` or `ASSERTS=1`, the asserts will
be there, but if `chpl` is ran with `--no-devel` then the asserts won't fire.
Further, if the asserts are there and `chpl` is ran with `--ignore-errors`, the
asserts will fire but any failures won't be fatal, just reported.

Added some unit tests to verify `CHPL_ASSERT` macro is appropriately controlled
by setting new assertion flags.

TESTING:

- [x] paratest
- [x] local build with `CHPL_DEVELOPER=1`
- [x] local build with `CHPL_DEVELOPER=0`
- [x] local build with `CHPL_DEVELOPER` unset
- [x] local build with `ASSERTS=0`
- [x] local build with `ASSERTS=1`
- [x] checks and fails asserts when built with `ASSERTS=1` and ran with `CHPL_DEVELOPER=1` or `--devel`
- [x] reports and ignores failures when built with `ASSERTS=1` and ran with `CHPL_DEVELOPER=1` or `--devel`, and `--ignore-errors`
- [x] skips checks when built with `ASSERTS=1` if ran with `--no-devel` or `CHPL_DEVELOPER=0` or `CHPL_DEVELOPER` unset
- [x] asserts always fire when running `make test-dyno`
- [x] performs no checks if built with `ASSERTS=0` and `CHPL_DEVELOPER=0` or unset, even when ran with `CHPL_DEVELOPER=1` or `--devel` later
- [x] performance testing indicates no appreciable difference compiling `hello.chpl` with `ASSERTS=0` vs. main with `ASSERTS=0`

Reviewed by @dlongnecke-cray - thanks!

### CODE REVIEW

File: frontend/include/chpl/framework/Context.h [hunk 1]
```diff
@@ -43,7 +43,12 @@
 namespace chpl {
   class Context;
 
-class ErrorBase;
+  // Whether or not to exit on assertion failure
+  extern bool assertionsAreFatal;
+  // Whether or not to enable assertions
+  extern bool assertionsAreOn;
```
Comment (mppf) [inline frontend/include/chpl/framework/Context.h:49]: I'd expect these to be declared in assertions.h / defined in assertions.cpp as well as the `set` calls for them

Comment (dlongnecke-cray) [reply frontend/include/chpl/framework/Context.h:49]: I also think if we have setters we should also have getters and not expose that these variables in the header as well

File: frontend/lib/util/assertions.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+               int lineno, const char* exprText) {
+  // assertions OFF: skip all checks
+  if (!assertionsAreOn) {
+    return;
+  }
+
+  // assertions ON: perform check, conditionally exit if assertionsAreFatal
+  if (!expr) {
+    // assertion failed: write out error
+    fprintf(stderr, "Assertion failed in [%s:%d] in %s: %s\n",
```
Comment (dlongnecke-cray) [inline frontend/lib/util/assertions.cpp:36]: I'd maybe change this message to: "[%s:%d in %s] Assertion failed: %s\n" but idk if it matters that much 😄.

File: compiler/main/driver.cpp [hunk 3]
```diff
@@ -1780,6 +1781,16 @@ static void validateSettings() {
   checkRuntimeBuilt();
 }
 
+static void setupDynoCompilerFlags() {
+  chpl::CompilerFlags compFlags = chpl::CompilerFlags();
+  compFlags.set(chpl::CompilerFlags::DEVELOPER, developer);
+  // TODO: Setting this flag causes a bunch of warnings to be printed
+  // but they are coming from our internal modules, so we should squelch them
+  // Seems to be coming from frontend/lib/uast/post-parse-checks.cpp:818
+  // compFlags.set(chpl::CompilerFlags::WARN_UNSTABLE, fWarnUnstable);
```
Comment (dlongnecke-cray) [inline compiler/main/driver.cpp:1790]: Yeah, we'll have to apply what I believe Lydia has been working on to silence warnings from internal/standard module code.

**Review by dlongnecke-cray — APPROVED** [review]
Looks great to me sans the change Michael suggested.

It might be good to add a unit test for the different flavors of the assertion.

File: compiler/main/driver.cpp [hunk 4]
```diff
@@ -1780,6 +1782,16 @@ static void validateSettings() {
   checkRuntimeBuilt();
 }
 
+static void setupDynoCompilerFlags() {
+  chpl::CompilerFlags compFlags = chpl::CompilerFlags();
+  compFlags.set(chpl::CompilerFlags::DEVELOPER, developer);
```
Comment (dlongnecke-cray) [inline compiler/main/driver.cpp:1787]: Do we need this flag for other work in the future? I don't see anything in dyno which uses it atm.

**Review by dlongnecke-cray — APPROVED** [review]
Let's also call `std::abort` instead of `exit(1)`

File: compiler/main/driver.cpp (hunk 4 above)
Comment (arezaii) [reply compiler/main/driver.cpp:1787]: We don't use it right now, I could remove it - although I expect we'd want it again at some point, maybe best to just wait until we have that use case.

**Pull request merged**